File: ai-service-python/app/extractors/canonical_resolver.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class CanonicalSignals:
    unique_key: bool = False
    alias: bool = False
    regex: bool = False
    embedding: float = 0.0
    code_graph_ref: bool = False
    llm_verify: bool = False


def _hit(value: bool) -> float:
    return 1.0 if value else 0.0


def _clamp(value: float | int | None) -> float:
    if value is None:
        return 0.0
    try:
        return max(0.0, min(1.0, float(value)))
    except (TypeError, ValueError):
        return 0.0


def score_breakdown(signals: CanonicalSignals) -> dict:
    unique_key = 0.40 * _hit(signals.unique_key)
    alias = 0.20 * _hit(signals.alias)
    regex = 0.15 * _hit(signals.regex)
    embedding = 0.10 * _clamp(signals.embedding)
    code_graph_ref = 0.10 * _hit(signals.code_graph_ref)
    llm_verify = 0.05 * _hit(signals.llm_verify)
    total = unique_key + alias + regex + embedding + code_graph_ref + llm_verify
    deterministic = any([unique_key, alias, regex])
    merge_allowed = total >= 0.85 and deterministic
    embedding_only = embedding > 0 and not any([unique_key, alias, regex, code_graph_ref, llm_verify])
    return {
        "uniqueKey": unique_key,
        "alias": alias,
        "regex": regex,
        "embedding": embedding,
        "codeGraphRef": code_graph_ref,
        "llmVerify": llm_verify,
        "total": total,
        "mergeAllowed": merge_allowed,
        "crossLinkHint": embedding_only and _clamp(signals.embedding) >= 0.85,
    }

File: backend-java/src/main/java/com/company/smartsupport/graph/CanonicalResolver.java
package com.company.smartsupport.graph;

import org.springframework.stereotype.Component;

import com.company.smartsupport.graph.dto.ScoreBreakdown;

@Component
public class CanonicalResolver {

    public static final double UNIQUE_KEY_WEIGHT = 0.40;
    public static final double ALIAS_WEIGHT = 0.20;
    public static final double REGEX_WEIGHT = 0.15;
    public static final double EMBEDDING_WEIGHT = 0.10;
    public static final double CODE_GRAPH_REF_WEIGHT = 0.10;
    public static final double LLM_VERIFY_WEIGHT = 0.05;
    public static final double MERGE_THRESHOLD = 0.85;

    public ScoreBreakdown score(CanonicalSignals signals) {
        double uniqueKey = hit(signals.uniqueKeyMatch()) * UNIQUE_KEY_WEIGHT;
        double alias = hit(signals.aliasMatch()) * ALIAS_WEIGHT;
        double regex = hit(signals.regexRuleMatch()) * REGEX_WEIGHT;
        double embedding = clamp(signals.embeddingSimilarity()) * EMBEDDING_WEIGHT;
        double codeGraphRef = hit(signals.codeGraphRefMatch()) * CODE_GRAPH_REF_WEIGHT;
        double llmVerify = hit(signals.llmVerified()) * LLM_VERIFY_WEIGHT;
        double total = uniqueKey + alias + regex + embedding + codeGraphRef + llmVerify;
        boolean strongDeterministicHit = uniqueKey > 0 || alias > 0 || regex > 0;
        boolean mergeAllowed = total >= MERGE_THRESHOLD && strongDeterministicHit;
        boolean embeddingOnly = embedding > 0 && uniqueKey == 0 && alias == 0 && regex == 0 && codeGraphRef == 0 && llmVerify == 0;
        boolean crossLinkHint = embeddingOnly && clamp(signals.embeddingSimilarity()) >= MERGE_THRESHOLD;
        return new ScoreBreakdown(uniqueKey, alias, regex, embedding, codeGraphRef, llmVerify, total, mergeAllowed, crossLinkHint);
    }

    private double hit(boolean value) {
        return value ? 1.0 : 0.0;
    }

    private double clamp(double value) {
        if (Double.isNaN(value) || value < 0) {
            return 0;
        }
        return Math.min(1, value);
    }

    public record CanonicalSignals(
            boolean uniqueKeyMatch,
            boolean aliasMatch,
            boolean regexRuleMatch,
            double embeddingSimilarity,
            boolean codeGraphRefMatch,
            boolean llmVerified) {
    }
}